compiler/damlc/daml-prim-src/GHC/Show/Text.daml

```diff
@@ -0,0 +1,127 @@
+primTextToCodePoints = primitive @"BETextToCodePoints"
+
+primIntToText : Int -> Text
+primIntToText = primitive @"BEToText"
+
+-- # Internal
+
+explode : Text -> [Chr]
+
+#ifdef SHOW_TEXT_CHR_NEWTYPE
+
+newtype Chr = MkChr Text
+  deriving (Eq, Ord)
+
+#define CH(t) (MkChr t)
+
+explode = map MkChr . primExplode
+
+#else
+
+type Chr = Text
+
+#define CH(t) t
+
+explode = primExplode
+
+#endif
+
+startsWithDigit : Text -> Bool
+startsWithDigit t = "0" <= t && t < ":" -- ':' immediately follows '9' in ASCII
+
+startsWithUppercaseH : Text -> Bool
+startsWithUppercaseH t = "H" <= t && t < "I"
+
+ord : Chr -> Int
+ord CH(t) = case primTextToCodePoints t of
+  [p] -> p
+  _ -> error $ "impossible: GHC.Show.Text.ord applied to argument with length /= 1, \"" ++ t ++ "\""
+
+type ShowS = Text -> Text
+
+showString : Text -> ShowS
+showString = (++)
+
+showInt : Int -> ShowS
+showInt = showString . primIntToText
+
+showLitText : Text -> ShowS
+showLitText = showLitChrs . explode
+
+showLitChrs : [Chr] -> ShowS
+showLitChrs []               = \s -> s
+showLitChrs (CH("\"") :: cs) = showString "\\\"" . showLitChrs cs
+showLitChrs (c        :: cs) = showLitChr c . showLitChrs cs
+
+showLitChr : Chr -> ShowS
+showLitChr c | c > CH("\DEL")       = showString "\\" . showInt (ord c) . protectEsc startsWithDigit
+showLitChr CH("\DEL")               = showString "\\DEL"
+showLitChr CH("\\")                 = showString "\\\\"
+showLitChr c@(CH(t)) | c >= CH(" ") = showString t
+showLitChr CH("\a")                 = showString "\\a"
+showLitChr CH("\b")                 = showString "\\b"
+showLitChr CH("\f")                 = showString "\\f"
+showLitChr CH("\n")                 = showString "\\n"
+showLitChr CH("\r")                 = showString "\\r"
+showLitChr CH("\t")                 = showString "\\t"
+showLitChr CH("\v")                 = showString "\\v"
+showLitChr CH("\SO")                = showString "\\SO" . protectEsc startsWithUppercaseH
+showLitChr c                        = showString "\\" . showString (ascii c)
+
+protectEsc : (Text -> Bool) -> ShowS
+protectEsc p = cont
+         where cont s | p s = showString "\\&" s
+               cont s       = s
+
+ascii : Chr -> Text
+ascii c = case c of
+  CH("\NUL") -> "NUL"; CH("\SOH") -> "SOH"; CH("\STX") -> "STX"; CH("\ETX") -> "ETX";
+  CH("\EOT") -> "EOT"; CH("\ENQ") -> "ENQ"; CH("\ACK") -> "ACK"; CH("\BEL") -> "BEL";
+
+  CH("\BS")  -> "BS";  CH("\HT")  -> "HT";  CH("\LF")  -> "LF";  CH("\VT")  -> "VT";
+  CH("\FF")  -> "FF";  CH("\CR")  -> "CR";  CH("\SO")  -> "SO";  CH("\SI")  -> "SI";
+
+  CH("\DLE") -> "DLE"; CH("\DC1") -> "DC1"; CH("\DC2") -> "DC2"; CH("\DC3") -> "DC3";
+  CH("\DC4") -> "DC4"; CH("\NAK") -> "NAK"; CH("\SYN") -> "SYN"; CH("\ETB") -> "ETB";
+
+  CH("\CAN") -> "CAN"; CH("\EM")  -> "EM";  CH("\SUB") -> "SUB"; CH("\ESC") -> "ESC";
+  CH("\FS")  -> "FS";  CH("\GS")  -> "GS";  CH("\RS")  -> "RS";  CH("\US")  -> "US";
+
+  CH("\SP")  -> "SP";
+
+  _ -> ""
```

Comment:
Did you copy this code from somewhere ? 
If yes, I think it would a good idea to add a comment. 

It is in any case a bit puzzling  ;). Why `\SO` is handled twice, why lower-case single character strings are handled differently form  upper-case multi-characters string

Reply:
Yes, it's adapted from GHC's `GHC.Show`. I've added a section explaining that and the changes needed here. 

Regarding `\SO` and the lower-case single-char vs upper-case multi-char control codes, the reason is that GHC's `asciiTab`, on which this is based, serves two purposes, first as a mapping from control character to the canonical representation, and second as the handler for `showLitChar`'s default case. In GHC, the special cases `\a\b\t\n\v\f\r` and `\SO` are handled directly by `showLitChr`, the former because the encoding differs, the latter because the possible ambiguity with `\SOH` must be considered. Since in Daml we don't need to expose the table, I've folded the cases for `\a\b\t\n\v\f\r` into `ascii` (now called `ctrlChrText`), and commented-out the redundant `\SO`.

Reply:
Thanks for the explanation. This make sens to me now.